rocksdb_admin/admin_handler.cpp

```diff
@@ -546,30 +551,26 @@ void AdminHandler::async_tm_addS3SstFilesToDB(
   // we couldn't afford to create a new client for every SST file downloading
   // request, which is not even on any critical code path. Otherwise, we will
   // see latency spike when uploading data to production clusters.
-  static const uint32_t s3_download_limit_mb = request->s3_download_limit_mb;
-  static const std::string s3_bucket = request->s3_bucket;
-  static auto s3Util = common::S3Util::BuildS3Util(
-    s3_download_limit_mb, s3_bucket);
+  RWSpinLock::UpgradedHolder upgraded_guard(s3_util_lock_);
+  if (s3_util_ == nullptr || should_new_s3_client(*s3_util_, request.get())) {
+    // Request with different ratelimit or bucket has to wait for old
+    // requests to drain.
+    RWSpinLock::WriteHolder write_guard(s3_util_lock_);
```

Comment:
shouldn't you upgrade from upgraded_guard?
Then the lock will be released in the if() block?